### PR TITLE
Self-supervised tutorial & update

### PR DESCRIPTION
**Pull request by sam1373**
- Basic tutorial for self-supervised pre-training and subsequent supervised fine-tuning
- Config for conformer SSL (may be changed later)
- Update to convert_to_tarred script to enable not having "text" field in manifest, and allow using "offset" for manifest without labels
- Add optional non-stride layers to reconstruction decoder
- Add option to load part of model from second nemo checkpoint from config (for example if we want to load encoder from one pre-trained model, and decoder from another)

### CODE REVIEW

File: nemo/core/classes/modelPT.py [hunk 1]
```diff
@@ -947,6 +947,33 @@ def maybe_init_from_pretrained_checkpoint(self, cfg: OmegaConf, map_location: st
 
                 del ckpt
 
+        if 'init_from_second_nemo_model' in cfg and cfg.init_from_second_nemo_model is not None:
+            with open_dict(cfg):
```
Comment (titu1994) [inline nemo/core/classes/modelPT.py:951]: Update doc string for this

Comment (titu1994) [inline nemo/core/classes/modelPT.py:951]: Change name. This seems confusing.

**Review by titu1994 — COMMENTED** [review]
Initial comments. Will review note book in seperate comment.

**Review by VahidooX — CHANGES_REQUESTED** [review]

File: examples/asr/conf/ssl/conformer/conformer_ssl.yaml [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+    n_fft: 512
+    log: true
+    frame_splicing: 1
+    dither: 0.00001
+    pad_to: 16
+    pad_value: 0.0
+
+  spec_augment:
+    _target_: nemo.collections.asr.modules.SpectrogramAugmentation
+    freq_masks: 4
```
[the remark this answers is not in the document]
Comment (sam1373) [reply examples/asr/conf/ssl/conformer/conformer_ssl.yaml:77]: I have found it to work well

File: nemo/core/classes/modelPT.py [hunk 3]
```diff
@@ -864,23 +894,37 @@ def maybe_init_from_pretrained_checkpoint(self, cfg: OmegaConf, map_location: st
...
+                include: List of strings, at least one of which needs to be contained in parameter name
+                to be loaded from this .nemo file
+
+                exclude: Optional list of strings, which can be used to exclude any parameter containing one of
+                these strings from being loaded from this .nemo file
 
             init_from_pretrained_model: Str name of a pretrained model checkpoint (obtained via cloud).
                 The model will be downloaded (or a cached copy will be used), instantiated and then
-                its state dict will be extracted.
+                its state dict will be extracted. If loading from multiple files, you can pass in a dict
```
[the remark this answers is not in the document]
Comment (sam1373) [reply nemo/core/classes/modelPT.py]: Added example

File: examples/asr/conf/ssl/conformer/conformer_ssl.yaml [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    feat_hidden: 128
+    feat_out: *dec_out
+    stride_layers: 0
+    non_stride_layers: 2
+
+
+  loss:
+    _target_: nemo.collections.asr.losses.ContrastiveLoss
+    in_dim: *n_mels
+    proj_dim: *dec_out
```
[the remark this answers is not in the document]
Comment (sam1373) [reply examples/asr/conf/ssl/conformer/conformer_ssl.yaml:129]: No, they have to have the same values, and yes it is just a placeholder.

**Review by VahidooX — APPROVED** [review]
LGTM!
